File: core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/worker/promisehandlers/BackpressureHandler.scala
package edu.uci.ics.amber.engine.architecture.worker.promisehandlers

import edu.uci.ics.amber.engine.architecture.worker.{
  BackpressurePause,
  DataProcessorRPCHandlerInitializer
}
import edu.uci.ics.amber.engine.architecture.worker.promisehandlers.BackpressureHandler.Backpressure
import edu.uci.ics.amber.engine.common.rpc.AsyncRPCServer.ControlCommand

object BackpressureHandler {
  final case class Backpressure(enableBackpressure: Boolean) extends ControlCommand[Unit]
}

/** Get queue and other resource usage of this worker
  *
  * possible sender: controller(by ControllerInitiateMonitoring)
  */
trait BackpressureHandler {
  this: DataProcessorRPCHandlerInitializer =>

  registerHandler { (msg: Backpressure, _) =>
    if (msg.enableBackpressure) {
      dp.pauseManager.pause(BackpressurePause)
    } else {
      dp.pauseManager.resume(BackpressurePause)
    }
  }

}

File: core/amber/src/main/python/core/architecture/handlers/backpressure_handler.py
from core.architecture.handlers.handler_base import Handler
from core.architecture.managers.context import Context
from core.architecture.managers.pause_manager import PauseType

from proto.edu.uci.ics.amber.engine.architecture.worker import BackpressureV2


class BackpressureHandler(Handler):
    cmd = BackpressureV2

    def __call__(self, context: Context, command: BackpressureV2, *args, **kwargs):
        if command.enable_backpressure:
            context.pause_manager.pause(PauseType.BACKPRESSURE_PAUSE)
        else:
            context.pause_manager.resume(PauseType.BACKPRESSURE_PAUSE)
